File: zero-decimal-currency-charged-100x/node/fix-zero-decimal-overcharge.js
/**
 * Find and refund WooCommerce orders in a zero decimal currency (JPY and friends)
 * that were charged 100x too much on Stripe.
 *
 * Stripe expects "amount" in the smallest unit of the currency. For two decimal
 * currencies like USD that is cents, so $50.00 is 5000. Zero decimal currencies such
 * as JPY, KRW, and VND have no smaller unit, so PY5000 is just 5000, not 500000. Code
 * that always multiplies the order total by 100 before sending it to Stripe overcharges
 * every zero decimal order by a factor of 100. This walks recent orders in the given
 * currencies, compares what Stripe actually charged to what the order should have cost,
 * and refunds the difference. Read only by default. Run once, or on a schedule.
 *
 * Guide: https://www.allanninal.dev/woocommerce/zero-decimal-currency-charged-100x/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 30);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

// https://docs.stripe.com/currencies#zero-decimal
const ZERO_DECIMAL_CURRENCIES = new Set([
  "bif", "clp", "djf", "gnf", "jpy", "kmf", "krw", "mga", "pyg",
  "rwf", "ugx", "vnd", "vuv", "xaf", "xof", "xpf",
]);

const PAID_STATUSES = new Set(["processing", "completed"]);

export function isZeroDecimal(currency) {
  return ZERO_DECIMAL_CURRENCIES.has((currency || "").toLowerCase());
}

/**
 * What Stripe's "amount" should be for this order total in this currency.
 *
 * Zero decimal currencies use the total as is (PY5000 -> 5000). Every other
 * currency uses the total times 100 (rounded) the usual way ($50.00 -> 5000).
 */
export function expectedMinorUnits(orderTotal, currency) {
  const total = parseFloat(orderTotal);
  if (isZeroDecimal(currency)) return Math.round(total);
  return Math.round(total * 100);
}

export function intentIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = order.transaction_id;
  return tid && tid.startsWith("pi_") ? tid : null;
}

/**
 * Pure decision: does this order need an overcharge refund, and for how much?
 *
 * Returns [action, reason, overchargeMinor]. overchargeMinor is the amount, in
 * the intent's own minor units, that should be refunded. It is 0 unless action
 * is "refund".
 */
export function decide(order, intent) {
  if (!PAID_STATUSES.has(order.status)) return ["skip", "order not in a paid state", 0];
  if (!isZeroDecimal(order.currency)) return ["skip", "not a zero decimal currency", 0];
  if (!intent) return ["skip", "no Stripe PaymentIntent on this order", 0];
  if (intent.status !== "succeeded") return ["skip", "Stripe payment did not succeed", 0];

  const charged = intent.amount_received || 0;
  const expected = expectedMinorUnits(order.total, order.currency);
  if (charged <= expected) return ["ok", "charge matches the order total", 0];

  // A 100x overcharge lands very close to charged / 100 == expected. Require
  // that ratio (within a small tolerance) so we only touch the bug this script
  // targets, not some unrelated pricing mismatch.
  if (expected <= 0 || Math.abs(charged - expected * 100) > Math.max(1, Math.floor(expected / 100))) {
    return ["mismatch", "overcharged but not by the 100x pattern", 0];
  }

  const alreadyRefunded = intent.amount_refunded || 0;
  const overcharge = charged - expected;
  const remaining = overcharge - alreadyRefunded;
  if (remaining <= 0) return ["ok", "overcharge already refunded", 0];

  return ["refund", "charged 100x the zero decimal total", remaining];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function getIntent(intentId) {
  if (!intentId) return null;
  try {
    return await stripe.paymentIntents.retrieve(intentId);
  } catch {
    return null;
  }
}

async function* paidOrders() {
  const after = new Date(Date.now() - LOOKBACK_DAYS * 86400000).toISOString();
  let page = 1;
  while (true) {
    const batch = await woo(`/orders?status=processing,completed&after=${after}&per_page=50&page=${page}`);
    if (!batch.length) return;
    for (const order of batch) yield order;
    page++;
  }
}

async function refundOvercharge(order, intent, overchargeMinor) {
  const chargeId = intent.latest_charge || intent.id;
  await stripe.refunds.create({
    ...(chargeId === intent.id ? { payment_intent: intent.id } : { charge: chargeId }),
    amount: overchargeMinor,
    reason: "duplicate",
    metadata: { reason: "zero_decimal_currency_100x_overcharge", order_id: String(order.id) },
  });
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Refunded a ${overchargeMinor} unit overcharge caused by treating ` +
            `${order.currency} as a two decimal currency. Stripe PaymentIntent ${intent.id}.`,
    }),
  });
}

export async function run() {
  let fixed = 0;
  for await (const order of paidOrders()) {
    const intent = await getIntent(intentIdOf(order));
    const [action, reason, overchargeMinor] = decide(order, intent);
    if (action === "mismatch") {
      console.warn(`Order ${order.id}: ${reason}`);
      continue;
    }
    if (action !== "refund") continue;
    console.warn(
      `Order ${order.id}: ${reason}. Overcharge is ${overchargeMinor} minor units. ` +
      `${DRY_RUN ? "would refund" : "refunding"}`
    );
    if (!DRY_RUN) await refundOvercharge(order, intent, overchargeMinor);
    fixed++;
  }
  console.log(`Done. ${fixed} order(s) ${DRY_RUN ? "to refund" : "refunded"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: zero-decimal-currency-charged-100x/python/fix_zero_decimal_overcharge.py
"""Find and refund WooCommerce orders in a zero decimal currency (JPY and friends)
that were charged 100x too much on Stripe.

Stripe expects "amount" in the smallest unit of the currency. For two decimal
currencies like USD that is cents, so $50.00 is 5000. Zero decimal currencies such
as JPY, KRW, and VND have no smaller unit, so PY5000 is just 5000, not 500000. Code
that always multiplies the order total by 100 before sending it to Stripe overcharges
every zero decimal order by a factor of 100. This walks recent orders in the given
currencies, compares what Stripe actually charged to what the order should have cost,
and refunds the difference. Read only by default. Run once, or on a schedule.
"""
import os
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("fix_zero_decimal_overcharge")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "30"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

# https://docs.stripe.com/currencies#zero-decimal
ZERO_DECIMAL_CURRENCIES = {
    "bif", "clp", "djf", "gnf", "jpy", "kmf", "krw", "mga", "pyg",
    "rwf", "ugx", "vnd", "vuv", "xaf", "xof", "xpf",
}

PAID_STATUSES = {"processing", "completed"}


def is_zero_decimal(currency):
    return (currency or "").lower() in ZERO_DECIMAL_CURRENCIES


def expected_minor_units(order_total, currency):
    """What Stripe's "amount" should be for this order total in this currency.

    Zero decimal currencies use the total as is (PY5000 -> 5000). Every other
    currency uses the total times 100 (rounded) the usual way ($50.00 -> 5000).
    """
    total = float(order_total)
    if is_zero_decimal(currency):
        return round(total)
    return round(total * 100)


def intent_id_of(order):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid if tid and tid.startswith("pi_") else None


def decide(order, intent):
    """Pure decision: does this order need an overcharge refund, and for how much?

    Returns a tuple of (action, reason, overcharge_minor). overcharge_minor is the
    amount, in the intent's own minor units, that should be refunded. It is 0 unless
    action is "refund".
    """
    if order["status"] not in PAID_STATUSES:
        return ("skip", "order not in a paid state", 0)
    if not is_zero_decimal(order.get("currency")):
        return ("skip", "not a zero decimal currency", 0)
    if intent is None:
        return ("skip", "no Stripe PaymentIntent on this order", 0)
    if intent.get("status") != "succeeded":
        return ("skip", "Stripe payment did not succeed", 0)

    charged = intent.get("amount_received", 0)
    expected = expected_minor_units(order["total"], order.get("currency"))
    if charged <= expected:
        return ("ok", "charge matches the order total", 0)

    # A 100x overcharge lands very close to charged / 100 == expected. Require
    # that ratio (within a small tolerance) so we only touch the bug this script
    # targets, not some unrelated pricing mismatch.
    if expected <= 0 or abs(charged - expected * 100) > max(1, expected // 100):
        return ("mismatch", "overcharged but not by the 100x pattern", 0)

    already_refunded = intent.get("amount_refunded", 0) or 0
    overcharge = charged - expected
    remaining = overcharge - already_refunded
    if remaining <= 0:
        return ("ok", "overcharge already refunded", 0)

    return ("refund", "charged 100x the zero decimal total", remaining)


def get_intent(intent_id):
    if not intent_id:
        return None
    try:
        return stripe.PaymentIntent.retrieve(intent_id)
    except stripe.error.InvalidRequestError:
        return None


def paid_orders():
    page = 1
    after = f"{__import__('datetime').date.today() - __import__('datetime').timedelta(days=LOOKBACK_DAYS)}T00:00:00"
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={"status": "processing,completed", "after": after, "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            yield order
        page += 1


def refund_overcharge(order, intent, overcharge_minor):
    charge_id = intent.get("latest_charge") or intent["id"]
    stripe.Refund.create(
        payment_intent=intent["id"] if charge_id == intent["id"] else None,
        charge=charge_id if charge_id != intent["id"] else None,
        amount=overcharge_minor,
        reason="duplicate",
        metadata={"reason": "zero_decimal_currency_100x_overcharge", "order_id": str(order["id"])},
    )
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={"note": f"Refunded a {overcharge_minor} unit overcharge caused by treating "
                      f"{order.get('currency')} as a two decimal currency. Stripe PaymentIntent "
                      f"{intent['id']}."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    fixed = 0
    for order in paid_orders():
        intent = get_intent(intent_id_of(order))
        action, reason, overcharge_minor = decide(order, intent)
        if action == "mismatch":
            log.warning("Order %s: %s", order["id"], reason)
            continue
        if action != "refund":
            continue
        log.warning(
            "Order %s: %s. Overcharge is %s minor units. %s",
            order["id"], reason, overcharge_minor, "would refund" if DRY_RUN else "refunding",
        )
        if not DRY_RUN:
            refund_overcharge(order, intent, overcharge_minor)
        fixed += 1
    log.info("Done. %d order(s) %s.", fixed, "to refund" if DRY_RUN else "refunded")


if __name__ == "__main__":
    run()
